File: content-solutions/queue-deque/queue-shuttle-bus.js
function solution(arrivals, buses, interval, capacity) {
  const answer = new Array(arrivals.length).fill(-1);
  const waiting = [];
  let head = 0;
  let nxt = 0;
  for (let bus = 1; bus <= buses; bus++) {
    const time = bus * interval;
    while (nxt < arrivals.length && arrivals[nxt] <= time) waiting.push(nxt++);
    const count = Math.min(capacity, waiting.length - head);
    for (let i = 0; i < count; i++) answer[waiting[head++]] = bus;
  }
  return answer;
}

File: content-solutions/queue-deque/queue-shuttle-bus.java
import java.util.*;

class Solution {
    public int[] solution(int[] arrivals, int buses, int interval, int capacity) {
        int[] answer = new int[arrivals.length];
        Arrays.fill(answer, -1);
        Deque<Integer> waiting = new ArrayDeque<>();
        int nxt = 0;
        for (int bus = 1; bus <= buses; bus++) {
            long time = (long) bus * interval;
            while (nxt < arrivals.length && arrivals[nxt] <= time) waiting.offer(nxt++);
            int count = Math.min(capacity, waiting.size());
            for (int i = 0; i < count; i++) answer[waiting.poll()] = bus;
        }
        return answer;
    }
}
